Doxygen/Docs/man/man3/Lights.3
.TH "Lights" 3 "Tue Dec 18 2012" "Version 9001" "OpenGL Flythrough" \" -*- nroff -*-
.ad l
.nh
.SH NAME
Lights \- 
.SH SYNOPSIS
.br
.PP
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBLights\fP (bool)"
.br
.RI "\fIDefault constructor\&. \fP"
.ti -1c
.RI "\fBLights\fP (GLuint &, bool complexSwitch)"
.br
.RI "\fIInitialization constructor\&. \fP"
.ti -1c
.RI "virtual \fB~Lights\fP ()"
.br
.RI "\fIDefault destructor\&. \fP"
.ti -1c
.RI "void \fBaddLightSource\fP (const \fBLightSource\fP &)"
.br
.RI "\fIAdd a light source to the list\&. \fP"
.ti -1c
.RI "void \fBremoveLightSource\fP (const unsigned int)"
.br
.RI "\fIRemoves the u'th element from the list of lights\&. \fP"
.ti -1c
.RI "void \fBremoveLastLightSource\fP ()"
.br
.RI "\fIRemoves the last light added to the container\&. \fP"
.ti -1c
.RI "void \fBmodifyLightSource\fP (const unsigned int u, const \fBLightSource\fP &l)"
.br
.RI "\fIChange an existing light source\&. \fP"
.ti -1c
.RI "void \fBinit_lights\fP (GLuint &)"
.br
.RI "\fIInitializes members in the object to link with shader uniform variables\&. \fP"
.ti -1c
.RI "unsigned int \fBgetNumLights\fP () const "
.br
.RI "\fIGet the number of lights we currently have active\&. \fP"
.ti -1c
.RI "bool \fBgetComplex\fP () const "
.br
.RI "\fIGet the lighting mode\&. \fP"
.ti -1c
.RI "void \fBsendAll\fP ()"
.br
.RI "\fIThis function call updates all the lighting arrays in the shader\&. \fP"
.in -1c
.SS "Private Attributes"

.in +1c
.ti -1c
.RI "vector< \fBLightSource\fP > \fBtheLights\fP"
.br
.ti -1c
.RI "float \fBambient\fP [MAX_LIGHTS *3]"
.br
.ti -1c
.RI "float \fBdiffuse\fP [MAX_LIGHTS *3]"
.br
.ti -1c
.RI "float \fBspecular\fP [MAX_LIGHTS *3]"
.br
.ti -1c
.RI "float \fBcolor\fP [MAX_LIGHTS *3]"
.br
.ti -1c
.RI "float \fBposition\fP [MAX_LIGHTS *3]"
.br
.ti -1c
.RI "float \fBdirection\fP [MAX_LIGHTS *3]"
.br
.ti -1c
.RI "GLuint \fBg_ambient\fP"
.br
.ti -1c
.RI "GLuint \fBg_diffuse\fP"
.br
.ti -1c
.RI "GLuint \fBg_specular\fP"
.br
.ti -1c
.RI "GLuint \fBg_color\fP"
.br
.ti -1c
.RI "GLuint \fBg_position\fP"
.br
.ti -1c
.RI "GLuint \fBg_direction\fP"
.br
.ti -1c
.RI "GLuint \fBg_numLights\fP"
.br
.ti -1c
.RI "const bool \fBcomplexSwitch\fP"
.br
.in -1c
.SH "Detailed Description"
.PP 
Definition at line 39 of file Lights\&.hpp\&.
.SH "Constructor & Destructor Documentation"
.PP 
.SS "Lights::Lights (GLuint &g_program, boolcSwitch)"

.PP
Initialization constructor\&. This constructor allows the implementor to forego a call to \fBinit_lights()\fP; If you can use this one, please do! 
.PP
\fBParameters:\fP
.RS 4
\fIg_program\fP A program object that we can link our arrays with\&. 
.RE
.PP

.PP
Definition at line 44 of file Lights\&.cpp\&.
.SH "Member Function Documentation"
.PP 
.SS "void Lights::addLightSource (const \fBLightSource\fP &l)"

.PP
Add a light source to the list\&. \fBParameters:\fP
.RS 4
\fIl\fP light source to add to the list\&. 
.RE
.PP

.PP
Definition at line 109 of file Lights\&.cpp\&.
.SS "void Lights::init_lights (GLuint &g_program)"

.PP
Initializes members in the object to link with shader uniform variables\&. IF YOU DON'T USE THE INITILIZATION CONSTRUCTOR, THIS MUST BE CALLED BEFORE TRYING TO CALL \fBsendAll()\fP; 
.PP
\fBParameters:\fP
.RS 4
\fIg_program\fP A gl program to use for linking with the shader\&. 
.RE
.PP

.PP
Definition at line 137 of file Lights\&.cpp\&.
.SS "void Lights::modifyLightSource (const unsigned intu, const \fBLightSource\fP &l)"

.PP
Change an existing light source\&. \fBParameters:\fP
.RS 4
\fIu\fP ONE INDEXED index to select a light source\&. 
.br
\fIl\fP light source to replace the existing light with\&. 
.RE
.PP

.PP
Definition at line 120 of file Lights\&.cpp\&.
.SS "void Lights::removeLightSource (const unsigned intu)"

.PP
Removes the u'th element from the list of lights\&. This function is 1-indexed\&. 
.PP
\fBParameters:\fP
.RS 4
\fIu\fP the index of the light to remove\&. 
.RE
.PP

.PP
Definition at line 87 of file Lights\&.cpp\&.
.SS "void Lights::sendAll ()"

.PP
This function call updates all the lighting arrays in the shader\&. If you make any changes to the lighting, you must call this function again\&. In fact, it probably belongs in your glutRedisplay() callback\&. 
.PP
Definition at line 177 of file Lights\&.cpp\&.

.SH "Author"
.PP 
Generated automatically by Doxygen for OpenGL Flythrough from the source code\&.


Classes/Lights.cpp
/**
   @file Light.hpp
   @author Nicholas StPierre
   @authors John Huston, Nicholas StPierre, Chris Compton
   @since 2012-11-20
   @date 2012-12-08
   @brief Implementation for the Light container class.
**/


#include <stdexcept>
#include <vector>
#include "vec.hpp"
#include "mat.hpp"
#include "LightSource.hpp"
#include "Lights.hpp"

using Angel::vec2 ;
using Angel::vec3 ;
using Angel::vec4 ; 
using Angel::mat4 ;

using std::vector ;
using std::exception ;
using std::out_of_range ;



/**
   Default constructor.
 **/
Lights::Lights(bool cSwitch) 
  : complexSwitch(cSwitch)
{

}

/**
   Initialization constructor.
   This constructor allows the implementor to forego a call to init_lights();
   If you can use this one, please do!
   @param g_program A program object that we can link our arrays with.
 **/
Lights::Lights( GLuint &g_program, bool cSwitch ) 
  : complexSwitch(cSwitch)
{

  if ( complexSwitch )
    {
    g_ambient  = glGetUniformLocation( g_program, "LightAmbientArray" ) ;
    g_diffuse  = glGetUniformLocation( g_program, "LightDiffuseArray" ) ;
    g_specular = glGetUniformLocation( g_program, "LightSpecularArray" ) ;
  
    } 
  else 
    {
       g_color = glGetUniformLocation( g_program, "LightColor" )  ;
    }


   g_position  = glGetUniformLocation( g_program, "LightPositionArray" ) ;
   g_direction = glGetUniformLocation( g_program, "LightDirectionArray" ) ;
   g_numLights = glGetUniformLocation( g_program, "numLights" ) ;

}

/**
   Default destructor.
 **/
Lights::~Lights(){

}

/**
   Get the lighting mode.
 **/
bool Lights::getComplex() const {
  return this->complexSwitch ;
}


/**
   Removes the u'th element from the list of lights.
   This function is 1-indexed.
   @param u the index of the light to remove.
 **/
void Lights::removeLightSource( const unsigned int u ){

  if ( u < theLights.size() )
    theLights.erase(theLights.begin() + (u-1) ) ;
  else throw out_of_range("Index out of bounds.");

}


/**
   Removes the last light added to the container.
 **/
void Lights::removeLastLightSource(){

  if ( getNumLights() > 0 ) theLights.pop_back() ;

}

/**
   Add a light source to the list.
   @param l light source to add to the list.
 **/
void Lights::addLightSource( const LightSource &l ){

  theLights.push_back( l ) ;

}

/**
   Change an existing light source.
   @param u ONE INDEXED index to select a light source.
   @param l light source to replace the existing light with.
 **/
void Lights::modifyLightSource( const unsigned int u, const LightSource &l ){

  if ( u <= theLights.size() && u != 0 )
    theLights[u-1] = l ;
  else throw out_of_range("Light index out of bounds.");


}

// void Lights::init_lights( GLuint &g_program, bool complexSwitch ) {

/**
   Initializes members in the object to link with shader uniform variables.
   IF YOU DON'T USE THE INITILIZATION CONSTRUCTOR,
   THIS MUST BE CALLED BEFORE TRYING TO CALL sendAll();
   @param g_program A gl program to use for linking with the shader.
 **/
void Lights::init_lights( GLuint &g_program ) {

  // To get the arrays to the shader, we are going to need handles to them.
  // These variables will hold the handles.

  // Note, we must use these names in the shader.
  if ( complexSwitch )
    {
      g_ambient  = glGetUniformLocation( g_program, "LightAmbientArray" ) ;
      g_diffuse  = glGetUniformLocation( g_program, "LightDiffuseArray" ) ;
      g_specular = glGetUniformLocation( g_program, "LightSpecularArray" ) ;  
    } 
  else 
    {
      g_color = glGetUniformLocation( g_program, "LightColor" )  ;
    }


   g_position  = glGetUniformLocation( g_program, "LightPositionArray" ) ;
   g_direction = glGetUniformLocation( g_program, "LightDirectionArray" ) ;
   g_numLights = glGetUniformLocation( g_program, "numLights" ) ;

  return ;

}

/**
   Get the number of lights we currently have active
 **/
unsigned int Lights::getNumLights() const {

  return (unsigned int) theLights.size();

}

/**
   This function call updates all the lighting arrays in the shader.
   If you make any changes to the lighting, you must call this function again.
   In fact, it probably belongs in your glutRedisplay() callback.
 **/
void Lights::sendAll() {

  unsigned int i ;
  vec4 temp4 ;

  for ( i = 0 ; i < theLights.size() && i < MAX_LIGHTS ; i++ ) {

    /*   color4 GetLight_specular() const;
	 color4 GetLight_diffuse() const;
	 color4 GetLight_ambient() const;
	 vec3 GetDirection() const;
	 vec4 GetPosition() const; */

    if (    complexSwitch    ) {
      temp4 = theLights[i].GetLight_ambient()  ;

      ambient[i*3]      =  temp4.x ; //r
      ambient[i*3 + 1]  =  temp4.y ; //g
      ambient[i*3 + 2]  =  temp4.z ; //b


      temp4 = theLights[i].GetLight_diffuse()  ;

      diffuse[i*3]    = temp4.x;
      diffuse[i*3 +1] = temp4.y;
      diffuse[i*3 +2] = temp4.z;


      temp4 = theLights[i].GetLight_specular()  ;

      specular[i*3]    = temp4.x;
      specular[i*3 +1] = temp4.y;
      specular[i*3 +2] = temp4.z;

    } else {

      temp4 = theLights[i].GetLight_color()  ;

      color[i*3]    = temp4.x;
      color[i*3 +1] = temp4.y;
      color[i*3 +2] = temp4.z;

    }

    temp4 =  theLights[i].GetPosition()    ;

    position[i*3]     = temp4.x ;
    position[i*3 +1]  = temp4.y ;
    position[i*3 +2]  = temp4.z ;


    temp4 =  theLights[i].GetDirection()    ;

    direction[i*3]    = temp4.x ;
    direction[i*3 +1] = temp4.y ;
    direction[i*3 +2] = temp4.z ;


  }

  // send the light color arrays

  if (  complexSwitch  ) {
    glUniform3fv( g_ambient,  i,  ambient  ) ;
    glUniform3fv( g_diffuse,  i,  diffuse  ) ;
    glUniform3fv( g_ambient,  i,  specular ) ;

  } else {

    glUniform3fv( g_color,  i,  color ) ;

  }

  // more light arrays: position information
  glUniform3fv( g_position ,  i, position  ) ;
  glUniform3fv( g_direction,  i, direction ) ;

  // send an integer to index all these lovely arrays
  glUniform1i( g_numLights, (GLint) this->getNumLights() ) ;

}

